torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+            else:
+                self._use_full_params()
+            # Prepare p.grad so that it is in the right shape, device, accumulated values, etc.
+            self._prep_grads_for_backward()
+
+            self.training_state = TrainingState.BACKWARD_PRE
+
+        def _register_hook(t: torch.Tensor) -> torch.Tensor:
+            if t.requires_grad:
+                t.register_hook(_pre_backward_hook)
+            return t
+
+        # Attach hooks to Tensor outputs.
+        outputs = _apply_to_tensors(_register_hook, outputs)
+
+        return outputs
+
+    def _register_post_backward_hooks(self) -> None:
+        """
+        Register backward hooks to reshard params and reduce-scatter grads.
+        This is called during forward pass. The goal is to attach a hook
+        on each of the parameter's gradient generating function (``grad_acc``
+        below) so that the hook is called *after* all gradients for that
+        param are computed.
+        Goals:
+        1. We want the hook to fire once and only once *after* all gradients
+        are accumulated for a param.
+        2. If it fires more than once, we end up incorrectly shard the grad
+        multiple times. (could lead to dimension too small)
+        3. If it fires once but too early or doesn't fire, we leave gradients
+        unsharded. (could lead to dimension too large)
+        Due to multiple-pass forward, this function can be called on
+        the same parameter multiple times in a single forward pass. If we register
+        the hook multiple time, we end up getting called multiple times. We
+        could try to get a new hook every time and delete the previous one
+        registered. However, due to *unknown reason* (I have debugged it for
+        a long time!), in mixed precision mode, we get two different ``grad_acc``
+        objects below during different calls of this function (in the same
+        forward pass). If we keep the last one, the hook end up firing too
+        early. In full precision mode, we luckily get the *same* ``grad_acc``
+        object, so deleting and re-registering still ensured the hook fire
+        once after all gradients are generated.
+        Empirically, keep the first hook register per forward pass seems to
+        work the best. We do need to remove the hook at the end of the
+        backward pass. Otherwise, the next forward pass will not register
+        a new hook, which is needed for a new forward pass.
+        """
+        if not torch.is_grad_enabled():
+            return  # don't register grad hooks if grad isn't enabled
+        for p in self.params:
+            if p.requires_grad:
+                if hasattr(p, "_shard_bwd_hook"):
+                    continue
+                # Register a hook on the first call, empirically, autograd
+                # fires it at the end for this param, which makes sense.
+                p_tmp = p.expand_as(p)  # Get a grad_fn on p_tmp.
```

Comment:
curious, why do we need `p.expand_as(p)` here? Is it possible to direct get access to the `AccumulateGrad` function?

And I assume we cannot use [`Tensor.register_hook()`](https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html) here because that might fire multiple times if a param is used multiple times? cc @albanD

Reply:
No the register_hook will be called once per call to backward even if the param is used multiple times.

Reply:
> No the register_hook will be called once per call to backward even if the param is used multiple times.

This means we can use `Tensor.register_hook()` here?

Reply:
Not if you allow autograd.grad: https://github.com/pytorch/pytorch/issues/35802

Reply:
p.expand_as(p) is used to get AccumulateGrad function here, we need to register post book on it, right? so that we can sync grads after all the accumulation is done

Reply:
> Not if you allow autograd.grad: #35802

@zhaojuanmao we do plan to support `autograd.grad` with FSDP?

> p.expand_as(p) is used to get AccumulateGrad function here, we need to register post book on it, right? so that we can sync grads after all the accumulation is done

Yep, if we don't plan to support `autograd.grad`, I assume `Tensor.register_hook()` can do the same thing without using the `AccumulateGrad` function?

Reply:
no plan to support autograd.grad() yet, because grad will be returned to users, not sure how the updated grads can be passed to post hooks. 

Still prefer to use AccumulateGrad function's post hooks, as those are used in DDP for a long time and approved to be robust

Reply:
We don't need to change anything in this PR. Just curious, what's the overhead of `p.expand_as(p)`? Would I be correct if I assume it won't create any new storage? cc @albanD

Reply:
based on the doc https://pytorch.org/docs/stable/generated/torch.Tensor.expand.html#torch.Tensor.expand, it will not create new memory, but return a new view on the existing tensor